lib/SILGen/SILGenConstructor.cpp

```diff
@@ -1204,9 +1204,45 @@ void SILGenFunction::emitClassConstructorInitializer(ConstructorDecl *ctor) {
     }
   }
 
+  CleanupStateRestorationScope SelfCleanupSave(Cleanups);
+
+  // If we're using a box for self, reload the value at the end of the init
+  // method.
+  auto cleanupLoc = CleanupLocation(ctor);
+  if (NeedsBoxForSelf) {
+    ManagedValue storedSelf =
+        ManagedValue::forBorrowedAddressRValue(VarLocs[selfDecl].value);
+    selfArg = B.createLoadCopy(cleanupLoc, storedSelf);
+  } else {
+    // We have to do a retain because we are returning the pointer +1.
+    //
+    // SEMANTIC ARC TODO: When the verifier is complete, we will need to
+    // change this to selfArg = B.emitCopyValueOperation(...). Currently due
+    // to the way that SILGen performs folding of copy_value, destroy_value,
+    // the returned selfArg may be deleted causing us to have a
+    // dead-pointer. Instead just use the old self value since we have a
+    // class.
+    selfArg = B.createCopyValue(cleanupLoc, selfArg);
+  }
```

Comment:
Naive idea to move the copy up a bit, but that's not correct since we end up

```
Begin Error in Function: '$s4main9SomeClassC4fromACyxGs7Decoder_p_tKcfc'
Error! Found a leak due to a consuming post-dominance failure!
Value:   %6 = copy_value %5 : $SomeClass<T>              // user: %44
Post Dominating Failure Blocks:
 bb6 bb5 bb4
End Error in Function: '$s4main9SomeClassC4fromACyxGs7Decoder_p_tKcfc'
Found ownership error?!
```

which I understand as not properly cleaning up the value? That makes sense... Doing a copy for all the distributed-sites also isn't quite right as the cleanups don't quite align... 

What would be the right away to model such thing so that the `emitDistributed...` things get the created copy and clean it up properly? Maybe @gottesmm you could have some hints?